File: Multiple Linear Regression.py
#Import Libraries.
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

#Importing the Dataset.
dataset = pd.read_csv('50_Startups.csv')
X = dataset.iloc[:, :-1].values
Y = dataset.iloc[:, 4].values

#Encoding Categorical Data.

#Encoding the Independent Variable.
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X = LabelEncoder()
X[:, 3] = labelencoder_X.fit_transform(X[:, 3])
onehotencoder = OneHotEncoder(categorical_features = [3])
X = onehotencoder.fit_transform(X).toarray()

#Avoiding the Dummy Variable Trap.
X = X[:, 1:]

#Encoding the Dependent Variable.
#labelencoder_y = LabelEncoder()
#y = labelencoder_y.fit_transform(y)

#Splitting data into training set and test set.
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

#Feature Scaling.
"""from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)"""

#Fitting Multiple Linear Regression to the Training Set.
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, y_train)

#Predicting the Test Set results.
y_pred = regressor.predict(X_test)

#Building the Optimal Model using Backward Elimination.
import statsmodels.formula.api as sm
X = np.append(arr = np.ones((50, 1)).astype(int), values = X, axis = 1)

File: Multiple Linear Regression.r
# Multiple Linear Regression.
# Importing the dataset.
dataset = read_csv('50_Startups.csv')
# dataset = dataset, [, 2:3]


# Encoding categorical data.
dataset$State = factor(dataset$State,
                        levels = c('New York', 'California', 'Florida'),
                        labels = c(1, 2, 3)) # corresponds with levels.


# Splitting data into training set and test set.
# install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(dataset$Profit, SplitRatio = 0.8)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)


# Feature scaling.
# training_set[, 2:3] = scale(training_set[, 2:3])
# test_set[, 2:3] = scale(test_set[, 2:3])


# Fitting multiple linear regression to training set.
regressor = lm(formula = Profit ~ .,
                data = training_set)


# Predicting the test set results.
y_pred = predict(regressor, newdata = test_set)


# Building the optimal model using backward elimination.
regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State,
                data = dataset)
summary(regressor)
regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend, # 'State' proven statistically insignificant; removed.
                data = dataset)
summary(regressor)
regressor = lm(formula = Profit ~ R.D.Spend + Marketing.Spend, # Removed 'Administration'.
                data = dataset)
summary(regressor)
regressor = lm(formula = Profit ~ R.D.Spend, # Removed 'marketing spend'.
                data = dataset)
summary(regressor)


# Automatic Backward Elimination.
backwardElimination <- function(x, sl) {
    numVars = length(x)
    for (i in c(1:numVars)){
      regressor = lm(formula = Profit ~ ., data = x)
      maxVar = max(coef(summary(regressor))[c(2:numVars), "Pr(>|t|)"])
      if (maxVar > sl){
        j = which(coef(summary(regressor))[c(2:numVars), "Pr(>|t|)"] == maxVar)
        x = x[, -j]
      }
      numVars = numVars - 1
    }
    return(summary(regressor))
  }

  SL = 0.05
  dataset = dataset[, c(1,2,3,4,5)]
  backwardElimination(training_set, SL)
